### PR TITLE
Use dbmap utils derive dump and open. Cleanup repetitions 

### PR DESCRIPTION
**Pull request by oxade**
* any test failures are strictly due to existing CI issues.

**This PR does two major things:**
1. Removes repetitive error-prone DBMap initializations and replaces them with `#[derive(DBMapUtils)]` which derives multiple features and allows configuring options such as `point_lookup` and `cache_capacity`. More options will be added as needed. Read more about this new feature [here].(https://github.com/MystenLabs/mysten-infra/pull/103)
2. Enables dumping all (but one) DBs, by leveraging the fact that `dump` is now auto derived. Secondarily, pagination is added to dump to limit resource exhaustion.

## 1. Macro use example:
For the index store we need to define a bunch of tables with different options and then open them as seen [here](https://github.com/MystenLabs/sui/blob/main/crates/sui-storage/src/indexes.rs#L23-L91)
The original implementation is below. Very repetitive and error-prone
**Before**
```
pub struct IndexStore {
    transactions_from_addr: DBMap<(SuiAddress, TxSequenceNumber), TransactionDigest>,

    transactions_to_addr: DBMap<(SuiAddress, TxSequenceNumber), TransactionDigest>,

    transactions_by_input_object_id: DBMap<(ObjectID, TxSequenceNumber), TransactionDigest>,

    transactions_by_mutated_object_id: DBMap<(ObjectID, TxSequenceNumber), TransactionDigest>,

    transactions_by_move_function:
        DBMap<(ObjectID, String, String, TxSequenceNumber), TransactionDigest>,

    timestamps: DBMap<TransactionDigest, u64>,
}

impl IndexStore {
    pub fn open<P: AsRef<Path>>(path: P, db_options: Option<Options>) -> Self {
        let (options, point_lookup) = default_db_options(db_options, Some(1_000_000));

        let db = {
            let path = &path;
            let db_options = Some(options.clone());
            let opt_cfs: &[(&str, &rocksdb::Options)] = &[
                ("transactions_from_addr", &options),
                ("transactions_to_addr", &options),
                ("transactions_by_input_object_id", &options),
                ("transactions_by_mutated_object_id", &options),
                ("transactions_by_move_function", &options),
                ("timestamps", &point_lookup),
            ];
            typed_store::rocks::open_cf_opts(path, db_options, opt_cfs)
        }
        .expect("Cannot open DB.");

        let (
            transactions_from_addr,
            transactions_to_addr,
            transactions_by_input_object_id,
            transactions_by_mutated_object_id,
            transactions_by_move_function,
            timestamps,
        ) = reopen!(
            &db,
            "transactions_from_addr"; <(SuiAddress, TxSequenceNumber), TransactionDigest>,
            "transactions_to_addr"; <(SuiAddress, TxSequenceNumber), TransactionDigest>,
            "transactions_by_input_object_id"; <(ObjectID, TxSequenceNumber), TransactionDigest>,
            "transactions_by_mutated_object_id"; <(ObjectID, TxSequenceNumber), TransactionDigest>,
            "transactions_by_move_function"; <(ObjectID, String, String, TxSequenceNumber), TransactionDigest>,
            "timestamps";<TransactionDigest, u64>
        );

        Self {
            transactions_from_addr,
            transactions_to_addr,
            transactions_by_input_object_id,
            transactions_by_mutated_object_id,
            transactions_by_move_function,
            timestamps,
        }
    }
```
**With the new logic, the options can be defined inline within the struct, and we don't need to implement the `open` feature as this is now done for us implicitly.
The code reduces to:** 

**after**
```
#[derive(DBMapUtils)]
pub struct IndexStore {
    #[options(cache_capacity = 1_000_000)]
    transactions_from_addr: DBMap<(SuiAddress, TxSequenceNumber), TransactionDigest>,

    #[options(cache_capacity = 1_000_000)]
    transactions_to_addr: DBMap<(SuiAddress, TxSequenceNumber), TransactionDigest>,

    #[options(cache_capacity = 1_000_000)]
    transactions_by_input_object_id: DBMap<(ObjectID, TxSequenceNumber), TransactionDigest>,

    #[options(cache_capacity = 1_000_000)]
    transactions_by_mutated_object_id: DBMap<(ObjectID, TxSequenceNumber), TransactionDigest>,

    #[options(cache_capacity = 1_000_000)]
    transactions_by_move_function:
        DBMap<(ObjectID, String, String, TxSequenceNumber), TransactionDigest>,

    #[options(optimization = "point_lookup", cache_capacity = 1_000_000)]
    timestamps: DBMap<TransactionDigest, u64>,
}
```

----------------------------------------------------------------------------------------------------------------------------

## 2. Any DB Dump Example
**Additionally since `dump` is now auto derived for all (but one -the WAL) DBs, we can expose this to the CLI.**

The following DBs can be dumped
```
    Validator,
    Gateway,
    Follower,
    Index,
    LocksService,
    NodeSync,
    Checkpoints,
```
*WAL DB is being fleshed out because the WAL can accept any type as its value so the code requires more context.

Example cmd for `NodeSync` `certs_and_fx` dump
```
/sui-tool db-tool --db-path ~/.sui/sui_config/authorities_db/495f3a436320b7c04ea67743523016337965059fee61ec27b8e6030d6b9745ee/node_sync_db/ dump --store NodeSync --table-name certs_and_fx --page-size 100 --page-num 0
```

### CODE REVIEW

File: crates/sui-core/Cargo.toml [hunk 1]
```diff
@@ -32,6 +32,17 @@ scopeguard = "1.1"
...
+syn = { version = "1.0.64", features = ["derive"] }
+quote = "1.0.9"
+proc-macro2 = "1.0.42"
+eyre = "0.6.8"
+thiserror = "1.0.31"
+collectable = "0.0.2"
+fdlimit = "0.2.1"
+
+
+
```
Comment (patrickkuo) [inline crates/sui-core/Cargo.toml:45]: these deps are not needed.

File: crates/sui-core/src/authority/authority_store_tables.rs [hunk 2]
```diff
@@ -110,225 +95,6 @@ pub struct StoreTables<S> {
 
     /// Map from each epoch ID to the epoch information. The epoch is either signed by this node,
     /// or is certified (signed by a quorum).
+    #[options(optimization = "point_lookup")]
     pub(crate) epochs: DBMap<EpochId, AuthenticatedEpoch>,
 }
-impl<S: Eq + Debug + Serialize + for<'de> Deserialize<'de>> StoreTables<S> {
```
Comment (patrickkuo) [inline crates/sui-core/src/authority/authority_store_tables.rs:115]: Nice!

File: crates/sui-tool/src/db_tool/mod.rs [hunk 3]
```diff
@@ -15,19 +15,32 @@ pub enum DbToolCommand {
...
-    #[clap(long = "table_name")]
+    #[clap(long = "table-name")]
     table_name: String,
+    /// The size of page to dump. This is a u16
+    #[clap(long = "page-size")]
+    page_size: u16,
+    /// The page number to dump
+    #[clap(long = "page-num")]
+    page_number: usize,
 }
```
Comment (patrickkuo) [inline crates/sui-tool/src/db_tool/mod.rs:32]: do we need to update docs for DB tools?

**Review by patrickkuo — APPROVED** [review]
Comments on unused dependencies and docs, otherwise LGTM, nice clean up!

File: crates/sui-tool/src/db_tool/mod.rs (hunk 3 above)
Comment (oxade) [reply crates/sui-tool/src/db_tool/mod.rs:32]: Good point. Will do as part of this issue
https://github.com/MystenLabs/sui/issues/3729

File: crates/sui-core/Cargo.toml (hunk 1 above)
Comment (oxade) [reply crates/sui-core/Cargo.toml]: Good catch. Removed them.

**Pull request merged**